### PR TITLE
llvm, halide, lit: migrate to Python 3.10

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

To dodge the versioned dependency conflict audit, I've added `coind3d`,
`gjs`, `pyside` and `pyside@2` to the allowlist. It's very unlikely that
these formulae use LLVM's Python bindings. LLVM is keg-only, and we
don't modify `PYTHONPATH`, so these formulae shouldn't even be able to
find LLVM's Python bindings. Also migrating these formulae would require
adding a ridiculous number of formulae to this PR. (cf. #87277)

However, just in case, I've also installed LLVM's Python bindings for
multiple versions of Python3. (This doesn't include LLDB, which has
linkage with the Python framework, which ties it to a specific Python
version.)

While we're here:
- remove duplication from the installed Xcode toolchain
- fix some version references so that they work for HEAD installs too

We clean up the duplication by installing the Xcode toolchain manually.

Closes #87483.

### CODE REVIEW

**Comment by carlocab** [comment]
```
==> brew audit --tap=homebrew/core --except=version
==> FAILED
watchman:
  * watchman contains conflicting version recursive dependencies:
      python@3.10, python@3.9
    View these with `brew deps --tree watchman`.
Error: 1 problem in 1 formula detected
Error: watchman contains conflicting version recursive dependencies:
  python@3.10, python@3.9
View these with `brew deps --tree watchman`.
```
What? No it doesn't. Not from this change, at least. Is the audit complaining about #87618? Ideas here, @Bo98?

**Comment by carlocab** [comment]
```
watchman
├── cmake  [build]
├── pkg-config  [build]
├── rust  [build]
│   ├── cmake  [build]
│   ├── ninja  [build]
│   │   └── python@3.10  [build] [test]
│   │       ├── pkg-config  [build]
│   │       ├── gdbm 
│   │       ├── mpdecimal 
│   │       ├── openssl@1.1 
│   │       │   └── ca-certificates 
│   │       ├── readline 
│   │       ├── sqlite 
│   │       │   └── readline 
│   │       └── xz 
│   ├── python@3.9  [build]
│   │   ├── pkg-config  [build]
│   │   ├── gdbm 
│   │   ├── mpdecimal 
│   │   ├── openssl@1.1 
│   │   │   └── ca-certificates 
│   │   ├── readline 
│   │   ├── sqlite 
│   │   │   └── readline 
│   │   └── xz 
│   ├── libssh2 
│   │   └── openssl@1.1 
│   │       └── ca-certificates 
│   ├── openssl@1.1 
│   │   └── ca-certificates 
│   └── pkg-config 
├── boost 
│   └── icu4c 
├── fmt 
│   └── cmake  [build]
├── folly 
│   ├── cmake  [build]
│   ├── pkg-config  [build]
│   ├── boost 
│   │   └── icu4c 
│   ├── double-conversion 
│   │   └── cmake  [build]
│   ├── fmt 
│   │   └── cmake  [build]
│   ├── gflags 
│   │   └── cmake  [build]
│   ├── glog 
│   │   ├── cmake  [build]
│   │   └── gflags 
│   │       └── cmake  [build]
│   ├── libevent 
│   │   ├── autoconf  [build]
│   │   │   └── m4 
│   │   ├── automake  [build]
│   │   │   └── autoconf 
│   │   │       └── m4 
│   │   ├── libtool  [build]
│   │   │   └── m4 
│   │   ├── pkg-config  [build]
│   │   └── openssl@1.1 
│   │       └── ca-certificates 
│   ├── lz4 
│   ├── openssl@1.1 
│   │   └── ca-certificates 
│   ├── snappy 
│   │   ├── cmake  [build]
│   │   └── pkg-config  [build]
│   ├── xz 
│   └── zstd 
│       └── cmake  [build]
├── gflags 
│   └── cmake  [build]
├── glog 
│   ├── cmake  [build]
│   └── gflags 
│       └── cmake  [build]
├── libevent 
│   ├── autoconf  [build]
│   │   └── m4 
│   ├── automake  [build]
│   │   └── autoconf 
│   │       └── m4 
│   ├── libtool  [build]
│   │   └── m4 
│   ├── pkg-config  [build]
│   └── openssl@1.1 
│       └── ca-certificates 
├── openssl@1.1 
│   └── ca-certificates 
├── pcre 
└── python@3.9 
    ├── pkg-config  [build]
    ├── gdbm 
    ├── mpdecimal 
    ├── openssl@1.1 
    │   └── ca-certificates 
    ├── readline 
    ├── sqlite 
    │   └── readline 
    └── xz 
```

**Comment by carlocab** [comment]
I'll just migrate `watchman` here too.

**Comment by carlocab** [comment]
Hmmm... I wonder if it's using the Linux dep tree to check for versioned conflicts.

**Comment by carlocab** [comment]
No, that doesn't seem to be true:
```
root@5c74b2189f14:/home/linuxbrew/.linuxbrew/Homebrew/Library/Taps/homebrew/homebrew-core# brew deps --tree --include-test --annotate watchman
watchman
├── boost
│   ├── icu4c
│   ├── bzip2
│   └── zlib
├── fmt
├── folly
│   ├── boost
│   │   ├── icu4c
│   │   ├── bzip2
│   │   └── zlib
│   ├── double-conversion
│   ├── fmt
│   ├── gflags
│   ├── glog
│   │   └── gflags
│   ├── libevent
│   │   └── openssl@1.1
│   │       └── ca-certificates
│   ├── lz4
│   ├── openssl@1.1
│   │   └── ca-certificates
│   ├── snappy
│   ├── xz
│   ├── zstd
│   │   └── zlib
│   └── gcc
│       ├── xz  [build] [test]
│       ├── gmp
│       │   └── xz  [build] [test]
│       ├── isl
│       │   ├── xz  [build] [test]
│       │   └── gmp
│       │       └── xz  [build] [test]
│       ├── libmpc
│       │   ├── gmp
│       │   │   └── xz  [build] [test]
│       │   └── mpfr
│       │       ├── xz  [build] [test]
│       │       └── gmp
│       │           └── xz  [build] [test]
│       ├── mpfr
│       │   ├── xz  [build] [test]
│       │   └── gmp
│       │       └── xz  [build] [test]
│       ├── zstd
│       │   └── zlib
│       ├── zlib
│       └── binutils
│           ├── xz  [build] [test]
│           └── zlib
├── gflags
├── glog
│   └── gflags
├── libevent
│   └── openssl@1.1
│       └── ca-certificates
├── openssl@1.1
│   └── ca-certificates
├── pcre
│   ├── bzip2
│   └── zlib
├── python@3.10
│   ├── gdbm
│   ├── mpdecimal
│   ├── openssl@1.1
│   │   └── ca-certificates
│   ├── readline
│   │   └── ncurses
│   ├── sqlite
│   │   ├── readline
│   │   │   └── ncurses
│   │   └── zlib
│   ├── xz
│   ├── bzip2
│   ├── expat
│   │   └── xz  [build] [test]
│   ├── libffi
│   ├── ncurses
│   ├── unzip
│   │   ├── zip  [test]
│   │   │   ├── bzip2
│   │   │   └── xz  [build] [test]
│   │   ├── bzip2
│   │   └── xz  [build] [test]
│   └── zlib
└── gcc
    ├── xz  [build] [test]
    ├── gmp
    │   └── xz  [build] [test]
    ├── isl
    │   ├── xz  [build] [test]
    │   └── gmp
    │       └── xz  [build] [test]
    ├── libmpc
    │   ├── gmp
    │   │   └── xz  [build] [test]
    │   └── mpfr
    │       ├── xz  [build] [test]
    │       └── gmp
    │           └── xz  [build] [test]
    ├── mpfr
    │   ├── xz  [build] [test]
    │   └── gmp
    │       └── xz  [build] [test]
    ├── zstd
    │   └── zlib
    ├── zlib
    └── binutils
        ├── xz  [build] [test]
        └── zlib
```

**Comment by Dayforce** [comment]
Good afternoon. I do not understand what this is about, because I am not familiar with this installation, but I really need to install llvm for another package to work. Explain what should I enter to make it work

**Comment by carlocab** [comment]
@Dayforce please post your question on the [discussions page](https://github.com/Homebrew/discussions/discussions). When you do, please include the output of `brew install llvm`, `brew config`, and `brew doctor`.

File: Formula/halide.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class Halide < Formula
   url "https://github.com/halide/Halide/archive/v13.0.0.tar.gz"
   sha256 "5ab7d5d9bc04ab0902a88751e535ff231ab9da1911e46f272a52cf41131609f3"
   license "MIT"
+  revision 1
```
Comment (cclauss) [inline Formula/halide.rb:7]: https://github.com/halide/Halide/releases
```suggestion
  url "https://github.com/halide/Halide/archive/v13.0.1.tar.gz"
  sha256 "TBD"
  license "MIT"
```

Comment (carlocab) [reply Formula/halide.rb:7]: That's for a separate PR. This one will be waiting a while for the self-hosted Linux runner.

Comment (cclauss) [reply Formula/halide.rb:7]: Please rebase to pick up #89931

**Comment by iMichka** [comment]
```
[34m==>[0m [1m/private/tmp/llvm-20211113-14370-cnghyh/llvm-project-13.0.0.src/llvm/stage1/bin/llvm-profdata merge -output=/private/tmp/llvm-20211113-14370-cnghyh/llvm-project-13.0.0.src/llvm/stage2/profiles/pgo_profile.prof 
....
/private/tmp/llvm-20211113-14370-cnghyh/llvm-project-13.0.0.src/llvm/stage2/profiles/default_15227703035553254730_0.profraw /private/tmp/llvm-20211113-14370-cnghyh/llvm-project-13.0.0.src/llvm/stage2/profiles/default_2152593734031737116_0.profraw[0m
2021-11-14T03:07:29.1578970Z warning: /private/tmp/llvm-20211113-14370-cnghyh/llvm-project-13.0.0.src/llvm/stage2/profiles/default_9261438557449877900_0.profraw: truncated profile data
2021-11-14T03:07:29.1580440Z error: no profile can be merged
```

Is this due to your toolchain changes? Do we need to ship these in this PR (with the risk to slow down the Python 3.10 migration).

**Comment by carlocab** [comment]
> Is this due to your toolchain changes? Do we need to ship these in this PR (with the risk to slow down the Python 3.10 migration).

No, I already had a successful run with my toolchain changes

**Comment by carlocab** [comment]
@iMichka The error was likely from `/tmp` not getting cleaned up properly. I was only waiting for a good time to re-run with the self-hosted runner. It didn't seem like a good time yet, but you seem to be in a hurry, so I re-ran it. It should be fine now, at least on macOS. I remember Linux being difficult here, but let's see.

**Comment by iMichka** [comment]
`halide` failed on Linux. It does not have a Linux bottle though, so it should be built on Linux. Or am I mistaken?

**Comment by carlocab** [comment]
> `halide` failed on Linux. It does not have a Linux bottle though, so it should be built on Linux. Or am I mistaken?

I don't understand your question. The Linux runner did attempt a build but ignored the build failure because halide does not have a bottle.

**Comment by iMichka** [comment]
> > `halide` failed on Linux. It does not have a Linux bottle though, so it should be built on Linux. Or am I mistaken?
> 
> I don't understand your question. The Linux runner did attempt a build but ignored the build failure because halide does not have a bottle.

Right, sorry. Got confused because the actual failure was the container failing to be deleted...

**Comment by carlocab** [comment]
Yea. Not sure what's going on there, but I've seen this error across multiple PRs. Help, @dawidd6?

**Comment by dawidd6** [comment]
Honestly I don't know what could cause that. It only happens on selfhosted runner.

**Comment by carlocab** [comment]
Can something else be cleaning it up before the `Stop containers` step is run?

**Comment by dawidd6** [comment]
Will try to investigate this.

**Review by iMichka — APPROVED** [review]

**Pull request closed without merging**